File: backend/src/models/submission.py
from datetime import datetime, timezone
from pydantic import BaseModel
from enum import Enum
from sqlmodel import Field, Session, SQLModel, create_engine, select

def utc_now(): return datetime.now(timezone.utc)

class SubmissionStatus(str, Enum):
    IN_QUEUE = "in queue"
    ACCEPTED = "accepted"
    WRONG_ANSWER = "wrong answer"
    TIME_LIMIT_EXCEEDED = "time limit exceeded"
    RUNTIME_ERROR = "runtime error"
    COMPILE_ERROR = "compile error"
    MEMORY_LIMIT_EXCEEDED = "memory limit exceeded"
    IDLENESS_LIMIT_EXCEEDED = "idleness limit exceeded"
    FAILED = "failed"

class Submission(SQLModel, table=True):
    __tablename__ = "submissions"
    id:int |None = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="users.id")
    problem_id: int = Field(foreign_key="problems.id")
    source_code: str
    submitted_at: datetime = Field(default_factory=utc_now)
    execution_time: int | None = None
    execution_memory: int | None = None 
    verdict : SubmissionStatus = Field(default=SubmissionStatus.IN_QUEUE)

File: betty-front/src/lib/types/submission.ts
export interface Submission {
  id: number;
  user_id: number;
  problem_id: number;
  source_code: string;
  submitted_at: string;
  execution_time: number | null;
  execution_memory: number | null;
  status: string | null;
  problem_name?: string;
  username?: string;
}

export interface SubmissionListItem {
  id: number;
  user_id: number;
  problem_id: number;
  problem_name: string;
  username: string;
  submitted_at: string;
  execution_time: number | null;
  execution_memory: number | null;
  status: string | null;
  language?: string;
}

export interface SubmitCodePayload {
  problem_id: number;
  source_code: string;
  language: string;
}

export type Verdict =
  | "Accepted"
  | "Wrong Answer"
  | "Time Limit Exceeded"
  | "Runtime Error"
  | "Compilation Error"
  | "Memory Limit Exceeded"
  | "Pending"
  | "Running";
